### ISSUE TITLE
Nautical Navigation Data

### ISSUE THREAD

**Issue by alex-w** [Stellarium/stellarium#408]
_Original report:_ https://bugs.launchpad.net/stellarium/+bug/1106745

This feature request from old bug tracker.

Object Data
 - [x] Rise, Meridian Passage and Set times
 - [x] GHA (Greenwich Hour Angle)
 - [x] SHA (Sidereal Hour Angle)
 - [x] LHA (Local Hour Angle)

Twilights (start & end times)
 - [x] civil
 - [ ] nautical
 - [ ] astronomical

_An additional proposal from the old bug tracker:_ indicate beginning and ending times of all twilights for the selected location in the sun data... also moon rise and set data.

**Comment by larawag** [comment]
Since this project seems to be still open, I tried to upload a first proposal for some nautical data like GHA, SHA, LHA, and UTC.

I would very appreciate, if you could give me any hints to optimize the code before I add some additional lines for the other topics.

**Comment by alex-w** [comment]
Seems UTC is not required here (plus user may switch time to UTC in any time). 

The twilight data may be added as extra info for the Sun.

**Comment by larawag** [comment]
Well, I would like to use stellarium to predict the azimuth and altitude of the navigational stars being visible during twilight. Then, I would like to compare the angles GHA and LHA with the values being listed in nautical almanacs. The angles in almanacs depend on UT1 and hence, the time measured on a boat are typically both zone time and UT1. As I expect that UT1 is not available in stellarium, I suggested to use instead UTC and show additionally zone time.

To make this comparison more comfortable, I thought to present all relevant nautical data simultaneously to the user.

Maybe the whole strategy can be added to the Astronomical Calculations Window?

**Comment by alex-w** [comment]
Yes, it's possible - just add a new tab into AstroCalc  (Navigation?) and put here calculation of all needed navigational data (include UT1 time).

**Comment by alex-w** [comment]
Of course Nav Stars plugin may have one additional tab in main GUI or separate dialog for navigational computations - it’s OK for me.

**Comment by gzotti** [comment]
Yes, the Nav stars plugin can have the GUI for it. But not the main program's AstroCalc GUI!

**Comment by alex-w** [comment]
Yes, this it perfect for me to use Nav Stars plugin for all navigational stuff.

**Comment by larawag** [comment]
I fully agree to your discussion and I will try to make a suggestion for a piece of code to update for the NavStar plugin. 

So, the idea to add only a few lines of code becomes a small project. As being a hobbyist (like most of us?) I ask you for some time to do that.

**Comment by gzotti** [comment]
Sure, take the time you require. Some issues have a long history :-) Most things are done by people who need some feature urgently themselves, because only these know particular inputs and outputs.

**Comment by alex-w** [comment]
An additional proposal from the old bug tracker: indicate beginning and ending times of all twilights for the selected location in the sun data... also moon rise and set data.

**Comment by larawag** [comment]
Hi, Andy, The attached picture shows parts of my current plugin extension as mentioned yesterday.

Thank you for sharing your files. Now we have the uncomfortable situation that we do have two different plugin drafts for the same task. Any suggestions how to proceed?

![NavStar_Twilight](https://user-images.githubusercontent.com/56893229/71785130-4b4fb380-2ffc-11ea-9d98-c57bff109bb5.png)

**Comment by A-j-K** [comment]
I already downloaded and compiled your code so I got a personal demonstration. Don't worry about two code repos. I'll halt what I'm doing on mine and we should proceed with yours, it's ahead of mine.

**Comment by A-j-K** [comment]
@larawag btw, I noticed that in your repo you have committed everything to the master branch. After reading the project Contrib guidelines I think we need to work in a branch and keep master clean so we can rebase (I see you haven't rebased since you forked). We need to fix these issues first.

**Comment by larawag** [comment]
@A-j-K Ok, now the files are dropped in [here](/larawag/stellarium/tree/issue408_navStars). Currently it is still a framework since it is WIP. 

You mentioned that you are going to write "educational packs" for students of Celestial Navigation. From my perspective, this seems to be a very big thing. Since CN is a navigation method that looks back on a long history, there are a lot of different methods to do CN. It feels hard to me, to decide which method  would be the best for implementing here, so I started the coding with the following targets:

- Enable the user to plan a trip (find the nav. stars for any location and the corresponding time in advance)
- Help students to compare their own computation with the results shown in the main window and the final result of CN
- Arouse curiosity about CN with this plugin

The sight reduction method I am going to use here is based on a least square approach. (So, it is not a semi-graphical method which might be a better choice for teaching purposes.)

**Comment by gzotti** [comment]
If you are going into historical methods, a description of Tobias Mayer's Lunar method would be esp. interesting, as this is much less well known. But only after the sight reduction...

**Comment by A-j-K** [comment]
@larawag My plan regarding "education packs" was to discuss with instructors and those who teach CN in a classroom setting how Stellarium could be used to enhance the understanding of CN. I realize many already have a set method of tuition and instruction but getting an expectation of "user experience" would be useful. So, more to follow on that after I have done some homework.

But I also like @gzotti idea of being able to go deeper and look at the history too. Most classes in CN focus on the "modern" sight reduction techniques as that's all they need to know to pass an exam. But I like the idea of being able to dive deeper into subjects like this. 

I'm just going to get your new branch and compile it up and then I'll get back to you.

**Comment by A-j-K** [comment]
@larawag Just a note, I ran in debug mode and on opening the config dialog got a runtime exception:-
NavStars.cpp line 112: Run-Time Check Failure #3 - The variable 'dayOfWeek' is being used without being initialized.

**Comment by A-j-K** [comment]
@larawag I've run into another issue, this one is somewhat more problematic and is concerned with the GHA (vernal equinox) ♈︎as displayed by ::addToExtraInfoString() function.

Steps to reproduce:-
1. Start your version of Stellarium.
2. Freeze the clock and do not allow Stellarium to go into real-time mode.
3. Set location 56d 26m 00s North and 3d 47m 15s West, 3m (this may not matter but I have this set)
4. Set the time to UTC 2020-01-13T07:12:34 (again, this may not matter but this is my setup)
5. Select Vega and note GHAa (I see 221d 36.8m)
6. Select Dubhe and note GHAa (I see 220d 23.2m)

With the Stellarium clock frozen I would not expect GHAa to move like this. My Almanac reduces to 220d 23.4m for this UTC time which appears ok for Dubhe but not Vega. Oddly, you can select objects all over and the split between these two GHAa values appears to be objects east of the local meridian or west.

**Comment by larawag** [comment]
@A-j-K Thank you for your hint. 

It seems to be more difficult since this unexpected behavior, i.e. the jump of the GHA value, happens at the meridian (what you reported) and - surprisingly - the jump is not restored 180° shifted to the meridian but its reset takes place at the autumnal equinox. I will take deeper look into this issue next weekend.

**Comment by larawag** [comment]
@A-j-K 
I could resolved this issue by modifying lines 227-230 in NavStars.cpp. Now you can find the current version in https://github.com/larawag/stellarium/tree/issue408_navStars 

I appreciate your attentive investigation of the plugin extension very much. - Thank you.

**Comment by A-j-K** [comment]
@larawag Thanks, I will have a look this evening when I get home from work. On a side note, I've managed to speak with folk who provide instruction on CN as part of overall marine navigation and have some feedback on how this may direct the UI experience. Again, I'll leave feedback this evening, I'm at work right now so can't do more than leave this reply :)

**Comment by A-j-K** [comment]
@larawag I just merged down your latest code. The GHA of the vernal equinox it now with 0.2' of an arc so it appears to be pretty close. However, your SHA is off by more than a 1' of arc and therefore subsequent GHA is also wrong. The DEC is with 0.3' of arc. See the attached image which shows your NavStars alongside my AstroNav plugin (which I verified with the nautical Almanac). Have a look at my code if it helps, I'm deriving from ERA (sidereal time).
![AstroNavstars2](https://user-images.githubusercontent.com/1134086/72570846-e9bdfd80-38b5-11ea-8fa7-4f4e821507c3.png)

**Comment by axd1967** [comment]
@A-j-K @larawag 
https://github.com/Stellarium/stellarium/pull/812 intended to add German nav stars (I didn't check the code to see what is meant by that, nor whether it has been merged).

If not merged, maybe that should be split off as a separate issue, independent of #955 (which added Russian and French nav stars), or added as a contribution to #955? A third approach is to add the German navstars to master, as #955 is still ongoing?

**Comment by gzotti** [comment]
#812 was apparently withdrawn silently. You can simply open a PR with the German navstars.

**Comment by alex-w** [comment]
I'll add German stars from #812 after merge Andy's branch

**Comment by larawag** [comment]
@A-j-K I really like your code.
@gzotti I closed #812 since I wanted to complete my NavStars plugin completely. Andy was now quicker...
@alex-w Thank you for adding German stars as well.

PS: I was absent for some weeks - obviously you have more spare time for coding than me ;-)

**Comment by A-j-K** [comment]
@larawag Hi, yeah, my kids have reached an age where I now have more spare time on my hands. I've been meaning for some time to work on Stellarium, I need a good hobby to take me towards retirement! (Well, that's still 15 years away but hey, I do like writing code and this project is definitely worth spending the time on, very enjoyable).

**Comment by larawag** [comment]
@A-j-K I fully agree to your opinion that contributing to Stellarium is a useful hobby. @xalioth indeed started a nice and useful project. 
BTW: Thanks expecially to @gzotti and @alex-w for maintaining the project.

**Comment by alex-w** [comment]
@A-j-K any news?

**Comment by alex-w** [comment]
v0.21.3:
[x] Twilight data

Twilight data was added for the Sun in version 0.21.3 (civil twilight by default)

**Comment by alex-w** [comment]
v0.22.2:

Twilights (start & end times)
 - [ ] civil
 - [ ] nautical
 - [ ] astronomical
